From bb0a3ac04f0bb25e1fd08740d2800f3c4b314d13 Mon Sep 17 00:00:00 2001
From: Aswin Murugan <quic_aswinm@quicinc.com>
Date: Wed, 2 Apr 2025 23:33:21 +0530
Subject: [PATCH] PENDING: Fix for UFS write issue

Encountered an issue with the MICRON Prod-MT128GAVAT2U31AA,Rev-0103 device
where the initial SCSI write operation consistently failed. Upon investigation,
it was determined that the problem caused from cache handling in the UFS device.
To address this, we disabled the write cache in the UFS by setting the FUA bit
in the second byte of the WRITE10 command.

Signed-off-by: Aswin Murugan <quic_aswinm@quicinc.com>
Upstream-Status: Pending
---
 drivers/scsi/scsi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/scsi/scsi.c b/drivers/scsi/scsi.c
index 8a177496bc..a297e269a5 100644
--- a/drivers/scsi/scsi.c
+++ b/drivers/scsi/scsi.c
@@ -108,7 +108,7 @@ static void scsi_setup_write_ext(struct scsi_cmd *pccb, lbaint_t start,
 				 unsigned short blocks)
 {
 	pccb->cmd[0] = SCSI_WRITE10;
-	pccb->cmd[1] = 0;
+	pccb->cmd[1] = 0x08; /* Set FUA bit to bypass write cache */
 	pccb->cmd[2] = (unsigned char)(start >> 24) & 0xff;
 	pccb->cmd[3] = (unsigned char)(start >> 16) & 0xff;
 	pccb->cmd[4] = (unsigned char)(start >> 8) & 0xff;
-- 
2.34.1

